components/sceneutil/workqueue.cpp

```diff
@@ -7,10 +7,10 @@ namespace SceneUtil
 
 void WorkItem::waitTillDone()
 {
+    OpenThreads::ScopedLock<OpenThreads::Mutex> lock(mMutex);
     if (mDone > 0)
```

Comment:
`mDone` appears to be an atomic variable, so the lock is unnecessary here. In the worst case, `mDone` is first checked to be 0, causing it to continue on, take the lock, and be checked as non-0 then release the lock and return. Taking the lock earlier doesn't really help, and can actually inhibit slight improvements... changing this first read to use acquire semantics would be faster for cases where the work item has been long done (relatively speaking). Only in cases where it was very recently completed might it "erroneously" fail, but in that case the lock and subsequent recheck would clear it all up.

As it is, with the lock moved like this you may as well just remove this first check since it checks the same thing again immediately with nothing in between. You only removed the pre-lock early-out test.

Reply:
Technically, since the doWork() is public, the `mDone` does not guarantee that the work item really does not do any job now.

Reply:
Can we somehow to make sure that we run every work item only once and only insode the WorkQueue, so the `mDone` is correct?

Reply:
I'm not sure what you mean. `mDone` is only ever set in `WorkItem::signalDone` from what I see. And given that `mDone` is an atomic variable, moving its early-out test under the lock doesn't improve anything since it was already safe to access (and instead just makes the test not early-out, and redundant with the following loop condition).

Reply:
There is nothing to prevent you from doing something like:

        mWorkItem = new CreateMapWorkItem(mWidth, mHeight, mMinX, mMinY, mMaxX, mMaxY, mCellSize, esmStore.get<ESM::Land>());
        mWorkQueue->addWorkItem(mWorkItem);
        mWorkItew->doWork();

In this case `mDone` just guarantees that both threads can update it without issues, It does not guarantee that access to task results is safe.

Also you can re-run already finished task manually. The `mDone` is still 1 in this case, despite the task is running and access to its results is not safe.

Reply:
Calling `doWork` manually after the item's been added to the queue doesn't look at all safe, unless all implementations of `doWork` are guaranteed thread-safe (and guaranteed to not return until the first caller is finished; which is up to the implementation, not this base). The caller has a responsibility to use an API correctly, and if `doWork` isn't thread-safe no one should be calling it in multiple threads.

In either case, `mDone` is only set to non-0 by a call to `signalDone`, which should only be called after `doWork` returns. As such, `waitTillDone` was fine as it was since it can't return before `signalDone` is called, which is after `doWork` is done.